### PR TITLE
Add Resources to metrics export APIs

### LINKED ISSUE #638: Per-instrument metric resources: OK or non-standard?
I was reflecting on how we handle Resources in the metrics pipelines and we haven't added support in the exporters yet. I noticed that we support resources on a per-instrument basis or fall-back to the SDK's resources. I wonder if this is "off spec"? I would note that Prometheus supports "constant labels" which is roughly the same, so maybe it's fine.

The question came about because I'm trying to figure out how to cache the encoding of a resource value in an exporter.

### PR DESCRIPTION
**Pull request by jmacd**
This adds a `*resource.Resource` argument to the `metric.Exporter` API.

It removes the per-instrument Resource feature, as discussed in #638. 
Resolves #638.

This adds simple resource printing to the stdout exporter. It leaves resources unimplemented for Prometheus and OTLP exporters. I'll file issues about that when this merges.

The motivation is that this is being requested in https://github.com/open-telemetry/opentelemetry-go-contrib/tree/master/exporters/metric/dogstatsd

I considered several ways of doing this. Currently Exporters have no direct reference to the SDK, which has the resources they want to use. Instead, the resource passes through the export pipeline and exporters are expected to cache an encoding of the resource. The resource must remain constant throughout the life of an exporter for a given SDK instance.

### CODE REVIEW

File: example/basic/main.go [hunk 1]
```diff
@@ -35,6 +36,7 @@ var (
 	barKey     = key.New("ex.com/bar")
 	lemonsKey  = key.New("ex.com/lemons")
 	anotherKey = key.New("ex.com/another")
+	resources  = resource.New(core.Key("rk1").String("rv11"), core.Key("rk2").Int64(5))
```
Comment (MrAlias) [inline example/basic/main.go:39]:
```suggestion
	resource  = resource.New(core.Key("rk1").String("rv11"), core.Key("rk2").Int64(5))
```

File: example/basic/main.go [hunk 2]
```diff
@@ -47,7 +49,7 @@ func initTracer() {
 	}
 	tp, err := sdktrace.NewProvider(sdktrace.WithSyncer(exp),
 		sdktrace.WithConfig(sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}),
-		sdktrace.WithResourceAttributes(core.Key("rk1").String("rv11"), core.Key("rk2").Int64(5)))
+		sdktrace.WithResource(resources))
```
Comment (MrAlias) [inline example/basic/main.go:52]:
```suggestion
		sdktrace.WithResource(resource))
```

File: exporters/metric/stdout/stdout.go [hunk 3]
```diff
@@ -130,26 +133,39 @@ func InstallNewPipeline(config Config) (*push.Controller, error) {
...
 	return pusher, nil
 }
 
-func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+func (e *Exporter) encodeResource(resource *resource.Resource) string {
+	if e.encodedResource != nil {
+		return *e.encodedResource
+	}
+	e.encodedResource = new(string)
+	resourceSlice := export.LabelSlice(resource.Attributes())
```
Comment (MrAlias) [inline exporters/metric/stdout/stdout.go:154]: It seems like this pattern makes [this addition](https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/resource/iterator.go) unneeded. We should probably remove it.

File: exporters/metric/stdout/stdout.go [hunk 4]
```diff
@@ -130,26 +133,39 @@ func InstallNewPipeline(config Config) (*push.Controller, error) {
...
 	batcher := ungrouped.New(selector, exporter.config.LabelEncoder, true)
-	pusher := push.New(batcher, exporter, period)
+	pusher := push.New(batcher, exporter, period, options...)
 	pusher.Start()
 
 	return pusher, nil
 }
 
-func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+func (e *Exporter) encodeResource(resource *resource.Resource) string {
```
Comment (MrAlias) [inline exporters/metric/stdout/stdout.go:149]: What happens if the exporter exports for two or more resources? As is the case when this exporter is registered with multiple controllers.

File: exporters/otlp/otlp.go [hunk 5]
```diff
@@ -210,7 +211,7 @@ func (e *Exporter) Stop() error {
 
 // Export implements the "go.opentelemetry.io/otel/sdk/export/metric".Exporter
 // interface. It transforms metric Records into OTLP Metrics and transmits them.
-func (e *Exporter) Export(ctx context.Context, cps metricsdk.CheckpointSet) error {
+func (e *Exporter) Export(ctx context.Context, _ *resource.Resource, cps metricsdk.CheckpointSet) error {
```
Comment (MrAlias) [inline exporters/otlp/otlp.go:214]: This is going to conflict with https://github.com/open-telemetry/opentelemetry-go/pull/626

File: exporters/metric/stdout/stdout.go (hunk 3 above)
Comment (jmacd) [reply exporters/metric/stdout/stdout.go:154]: Yeah -- I was wishing for a more consolidated approach to Resource and metric.Labels. I'll mention it in a TODO.

File: exporters/metric/stdout/stdout.go (hunk 4 above)
Comment (jmacd) [reply exporters/metric/stdout/stdout.go:149]: I guess I was hoping no one would say this. :)

OTOH, that was my motivation for adding the comment about immutability and using pointers, so I'll add a map of resource to encoded value here.

**Comment by krnowak** [comment]
Looks good, but needs rebase.

**Pull request closed without merging**